### ISSUE TITLE
Information about golang versions

### ISSUE THREAD

**Issue by juev** [kevincobain2000/gobrew#199]
Now we independently generate JSON with available versions of Golang in order to have information about the versions of Golang available for installation every time we start gobrew.

That is, we need to have a separate branch in the repository to store the result, we need to have automation in which we make a request to the Golang repository to collect information on all tags and form the Json file we need, which we update in the branch if it is updated.

There is another source of information on Golang versions. This is the official repository:

```plain
https://go.googlesource.com/go/+refs?format=JSON
```

Adding the suffix `?format=JSON` allows you to get the result in Json format. The result contains all existing development branches, including tags. Which we use in our work.

I tested requests for the specified url, found that there are no restrictions on requests. At least 60 requests per minute can withstand without problems and does not start blocking requests. Thus, we can switch to using this url as a source of information for Gobrew.

The only thing is that we will also need to leave the existing automation for some more time in order for the old versions of gobrew to continue to work without problems. In the future, it will be possible to disable automation, leaving the file available in the repository. But I think it can be discussed separately.

There is also an idea to cache requests to the server `go.googlesource.com` in the system as a local file. That is, we fly the request, then either save the full answer on the file system and then parse the result at each run. Or we immediately parse and save the finished result on the file system. We also save the request time. With each subsequent launch of gobrew, we use the file cache to obtain information about available versions of Golang. And for example, once an hour or even once a day, we update the file cache the next time we start gobrew.

We can also provide an additional gobrew launch key, which will lead to an accelerated update of the file cache.

Together with the changes, we get rid of the custom storage of information on available versions of Golang. We will always have up-to-date information on versions. And the file cache will allow minimal access to Google servers.

All these are thoughts to think about and to try to implement. All this is negotiable.

And I'll be glad to hear your opinion @kevincobain2000

**Comment by kevincobain2000** [comment]
60 rpm is not bad, accompanied with a short lived cache would do. 
We run about 20 ci jobs per minute on a busy day of development.

**Comment by juev** [comment]
Then I'll try to implement it as much as I'm a little freer.

**Comment by kevincobain2000** [comment]
Take it easy bud.

**Comment by juev** [comment]
I didn't think about something at first, only now I looked at the amount of data that I have to download from us and from the Google server.

<img width="319" alt="Screenshot 2024-07-30 at 09 48 51" src="https://github.com/user-attachments/assets/f1b12266-7a07-47a2-9eef-3c71507e10c2">

That is, it's not an option.

I'll think about the need for a file cache with a limited time.

And then I thought that if we don't need to check versions, and just set the version, then we don't need to ask for Json, handle errors while downloading the version. The only question is the appropriateness of this.

**Comment by juev** [comment]
Yes, also, why do we store so many fields in Json?

```json
  {
    "ref": "refs/tags/go1",
    "node_id": "MDM6UmVmMjMwOTY5NTk6cmVmcy90YWdzL2dvMQ==",
    "url": "https://api.github.com/repos/golang/go/git/refs/tags/go1",
    "object": {
      "sha": "6174b5e21e73714c63061e66efdbe180e1c5491d",
      "type": "commit",
      "url": "https://api.github.com/repos/golang/go/git/commits/6174b5e21e73714c63061e66efdbe180e1c5491d"
    }
  },
```

We only use ref in gobrew, don't we?

Perhaps it is worth storing just an array of versions, immediately choosing the right ones from all. For example, we don't need such `weekly.2010-07-01`. That is, immediately filter and select the necessary versions and store only them. To gobrew on the side, just pick up and immediately use without processing.

You can also take automation to a separate repository and publish the file as a static site, something like `gobrew-json.github.io/golang-tags.json`.

What do you think?

**Comment by kevincobain2000** [comment]
Jeez, 25 MB.

>We only use ref in gobrew, don't we?

Yes. I remember, we needed to push that solution fast. At the same time, I was also trying to use `gobrew` for actions as well. So dumped everything inside the JSON.

> Perhaps it is worth storing just an array of versions,

Yea, just the versions array would make sense.

> a separate repository and publish the file as a static site

This too, both you and I discussed it in the past and then decided to keep it simple and all in one place. I mean, the branch is protected and all. We all good.
